includes/Abstracts/Abstract_Ability.php

```
@@ -97,4 +100,81 @@ abstract protected function permission_callback( $input );
+	}
+
+	/**
+	 * Loads system instruction from a PHP file in the feature's directory.
+	 *
+	 * Automatic detection order:
+	 * 1. `system-instruction.php`
+	 * 2. `prompt.php`
+	 *
+	 * PHP files should return a string directly, e.g.:
+	 * ```php
+	 * <?php
+	 * return 'Your system instruction text here...';
+	 * ```
+	 *
+	 * @since 0.1.0
+	 *
+	 * @param string|null $filename Optional. Explicit filename to load. If not provided,
+	 *                              attempts to load `system-instruction.php` or `prompt.php`.
+	 * @return string The contents of the file, or empty string if file not found.
+	 */
+	protected function load_system_instruction_from_file( ?string $filename = null ): string {
+		// Get the feature's directory using reflection.
+		$reflection = new ReflectionClass( $this );
+		$file_name  = $reflection->getFileName();
+
+		if ( ! $file_name ) {
+			return '';
+		}
+
+		$feature_dir = dirname( $file_name );
+
+		// If explicit filename provided, use it.
+		if ( null !== $filename ) {
+			$file_path = trailingslashit( $feature_dir ) . $filename;
+
+			if ( file_exists( $file_path ) && is_readable( $file_path ) ) {
+				// PHP files should return a string directly.
+				$content = require_once $file_path; // phpcs:ignore WordPressVIPMinimum.Files.IncludingFile.UsingVariable
+
+				return is_string( $content ) ? wp_strip_all_tags( $content ) : '';
+			}
+
+			return '';
+		}
+
+		// Automatic detection if no filename provided.
+		$possible_files = array(
+			'system-instruction.php',
+			'prompt.php',
+		);
+
+		foreach ( $possible_files as $possible_file ) {
+			$file_path = trailingslashit( $feature_dir ) . $possible_file;
+
+			if ( file_exists( $file_path ) && is_readable( $file_path ) ) {
+				// PHP files should return a string directly.
+				$content = require $file_path; // phpcs:ignore WordPressVIPMinimum.Files.IncludingFile.UsingVariable
+
+				return is_string( $content ) ? wp_strip_all_tags( $content ) : '';
+			}
+		}
+
+		return '';
```

Do we want silent failure here? Maybe yes, just double-checking. 🙂

---

That was the thought, yes, as the Prompt Builder handles that fine. Could change this to return an error that we output, though that would require us to not have the same chaining we have right now on the Prompt Builder as we'd need to check if an error was returned before calling `using_system_instruction`